api/server/router/network/network_routes.go

```diff
@@ -34,16 +35,29 @@ func (n *networkRouter) getNetworksList(ctx context.Context, w http.ResponseWrit
 
 	// Combine the network list returned by Docker daemon if it is not already
 	// returned by the cluster manager
+	var wg sync.WaitGroup
+	networkResourceChannel := make(chan types.NetworkResource, 1)
 SKIP:
 	for _, nw := range n.backend.GetNetworks() {
 		for _, nl := range list {
 			if nl.ID == nw.ID() {
 				continue SKIP
 			}
 		}
-		list = append(list, *n.buildNetworkResource(nw))
+		wg.Add(1)
+		go func(tmpNw libnetwork.Network) {
+			networkResourceChannel <- *n.buildNetworkResource(tmpNw)
+		}(nw)
 	}
 
+	go func() {
+		for resource := range networkResourceChannel {
+			list = append(list, resource)
```

Comment:
`networkResourceChannel ` is a channel which buffer size is 1. Then the last `goroutine` will only set one value  into channel at the same time.  This routine will read this channel synchronously, it won't cause some race condition. @unclejack